### PR TITLE
openssl_*: improve passphrase handling for private keys in PyOpenSSL

### LINKED ISSUE #53376: openssl_* modules: PyOpenSSL asks user for password if trying to open password-protected key
##### SUMMARY
As reported in #53360, the `openssl_certificate` module will hang (on remote nodes) if a password-protected private key is used, while the `passphrase` option is not provided to the module. If the module is run on `localhost` (with `local` connection), one can see that PyOpenSSL asks for a password: it shows a `Enter PEM pass phrase:` prompt and waits for the user to enter something before trying to open the private key.

This problem also appears with the `openssl_csr` module and with `openssl_privatekey` itself (in all cases with the `PyOpenSSL` backend, with the `cryptography` backend this doesn't happen).

CC @johnwesley

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
openssl_privatekey
openssl_csr
openssl_certificate

##### ANSIBLE VERSION
```paste below
2.7.x
```

##### STEPS TO REPRODUCE
```yml
---
- hosts: localhost
  tasks:
  - openssl_privatekey:
      path: bla.key
      passphrase: xyz
      cipher: aes256
      select_crypto_backend: pyopenssl
  - openssl_privatekey:
      path: bla.key
      # passphrase: xyz
      # cipher: aes256
      select_crypto_backend: pyopenssl
  - openssl_csr:
      path: bla.csr
      privatekey_path: bla.key
      # privatekey_passphrase: xyz
      select_crypto_backend: pyopenssl
  - openssl_certificate:
      path: bla.pem
      privatekey_path: bla.key
      # privatekey_passphrase: xyz
      csr_path: bla.csr
      provider: selfsigned
```
This playbook hangs on every step (except the first). Uncommenting the options in each step make the playbook finish this step (and hang on the next one where the passphrase is still commented out).

(The `select_crypto_backend: pyopenssl` option is only needed for `devel` branch; for 2.7.x it must be left away.)

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Fixes #53376. Currently just experimenting.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
openssl_privatekey
openssl_csr
openssl_certificate
openssl_pkcs12

### CODE REVIEW

**Comment by felixfontein** [comment]
I think this just needs some tests, and then it should be fine. As long as the behavior of strictly checking the passphrase (i.e. fail if passphrase is provided, but the key is not protected by one) is OK.

**Comment by felixfontein** [comment]
ready_for_review

@Shaps: this might create conflicts when you start adding a `cryptography` backend to `openssl_certificate`, so maybe we should get it merged first. What do you think?

**Comment by Shaps** [comment]
@felixfontein yeah, it will create conflicts with the new backend. If this doesn't take too long to get merged I'll wait for it, otherwise fix the conflicts later

File: lib/ansible/module_utils/crypto.py [hunk 1]
```diff
@@ -78,20 +82,46 @@ def get_fingerprint(path, passphrase=None):
...
+            # a passphrase (in case we did try the empty string before, make
+            # sure that the key is not protected by the empty string)
+            try:
+                crypto.load_privatekey(crypto.FILETYPE_PEM,
+                                       priv_key_detail,
+                                       to_bytes('y' if passphrase == 'x' else 'x'))
+                if passphrase is not None:
+                    # Since we can load the key without an exception, the
+                    # key isn't password-protected
+                    raise OpenSSLBadPassphraseError('Passphrase provided, but private key is not password-protected!')
```
Comment (Shaps) [inline lib/ansible/module_utils/crypto.py:118]: Should we really raise an error if the key is not password protected? Wouldn't a warning be enough here?

**Comment by ffdybuster** [comment]
@Shaps no idea how long it will take, it needs reviews first ;) I think the most controversial change is that if you specify a passphrase while the private key doesn't need one, it will now fail. Though it already does that for the `cryptography` backends. The main question is whether we want this behavior. @Spredzy @MarkusTeufelberger any opinions on this?

File: lib/ansible/module_utils/crypto.py (hunk 1 above)
Comment (ghost) [reply lib/ansible/module_utils/crypto.py:118]: The openssl_privatekey needs this exception to know that it needs to regenerate the private key. For the other modules, it's a good question (see my comment below)...

File: lib/ansible/modules/crypto/openssl_certificate.py [hunk 2]
```diff
@@ -1002,10 +1011,15 @@ def generate(self, module):
 
         self.assertonly()
 
-        if self.privatekey_path and \
-           not super(AssertOnlyCertificate, self).check(module, perms_required=False):
+        try:
+            if self.privatekey_path and \
+               not super(AssertOnlyCertificate, self).check(module, perms_required=False):
+                self.message.append(
+                    'Certificate %s and private key %s does not match' % (self.path, self.privatekey_path)
```
Comment (Shaps) [inline lib/ansible/modules/crypto/openssl_certificate.py]:
```suggestion
                    'Certificate %s and private key %s do not match' % (self.path, self.privatekey_path)
```

**Comment by Shaps** [comment]
shipit

**Comment by felixfontein** [comment]
We've decided to merge this early, and move the discussion into a new issue (just created #53535 for that).

**Comment by gundalow** [comment]
Given above, merged. Thanks all

**Comment by felixfontein** [comment]
@Shaps thanks for reviewing!
@gundalow thanks for merging!

**Pull request merged**